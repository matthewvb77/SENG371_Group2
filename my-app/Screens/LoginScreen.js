import React, { useState, useEffect } from "react";
import {
  View,
  Text,
  TextInput,
  Button,
  TouchableOpacity,
  Image,
} from "react-native";
import {
  auth,
  db,
  doc,
  getDoc,
  signInWithEmailAndPassword,
  GoogleAuthProvider,
} from "../firebase";
import * as WebBrowser from "expo-web-browser";
import * as Crypto from "expo-crypto";
import {
  makeRedirectUri,
  useAuthRequest,
  ResponseType,
} from "expo-auth-session";

WebBrowser.maybeCompleteAuthSession();

const LoginScreen = ({ navigation }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");

  const redirectUri = makeRedirectUri({ useProxy: true });

  const [request, response, promptAsync] = useAuthRequest(
    {
      responseType: ResponseType.Token,
      clientId:
        "162824721551-inb5r4haq77q7k5rf7a2acorgppbv951.apps.googleusercontent.com",
      scopes: ["openid", "profile", "email"],
      redirectUri: redirectUri,
      codeVerifier: "",
    },
    {
      authorizationEndpoint: "https://accounts.google.com/o/oauth2/v2/auth",
      tokenEndpoint: "https://oauth2.googleapis.com/token",
      revocationEndpoint: "https://oauth2.googleapis.com/revoke",
    }
  );

  useEffect(() => {
    if (response?.type === "success") {
      const { access_token } = response.params;

      const credential = auth.GoogleAuthProvider.credential(access_token);

      auth
        .signInWithCredential(credential)
        .then((userCredentials) => {
          const uid = userCredentials.user.uid;
          getDoc(doc(db, "users", uid)).then((docSnap) => {
            if (docSnap.exists()) {
              const user = docSnap.data();
              navigation.navigate("Dashboard", { user: user });
            } else {
              console.log("No such document");
            }
          });
        })
        .catch((error) => {
          alert(error);
        });
    }
  }, [response]);

  const handleLoginPress = () => {
    signInWithEmailAndPassword(auth, email, password)
      .then((userCredentials) => {
        const uid = userCredentials.user.uid;
        getDoc(doc(db, "users", uid)).then((docSnap) => {
          if (docSnap.exists()) {
            const user = docSnap.data();
            navigation.navigate("Dashboard", { user: user });
          } else {
            console.log("No such document");
          }
        });
      })
      .catch((error) => {
        alert(error);
      });
  };

  const handleGoogleLoginPress = async () => {
    const state = await Crypto.getRandomBytesAsync(6).then((bytes) => {
      return bytes.reduce(
        (str, byte) => str + byte.toString(16).padStart(2, "0"),
        ""
      );
    });

    const authUrl = `${request.url}&state=${state}`;

    promptAsync({ authUrl });
  };

  return (
    <View className='flex-1 pt-10'>
      <View className='flex items-center'>
        <Text className='text-4xl font-bold mt-12'>Login</Text>
      </View>
      <View className='mt-12 px-5'>
        <TextInput
          className='h-12 px-4 bg-gray-100 border border-gray-500 rounded-lg mb-4'
          placeholder='Email'
          onChangeText={(value) => setEmail(value)}
        />
        <TextInput
          className='h-12 px-4 bg-gray-100 border border-gray-500 rounded-lg mb-4'
          placeholder='Password'
          onChangeText={(value) => setPassword(value)}
        />
        <TouchableOpacity
          onPress={handleLoginPress}
          className='items-center bg-blue-500 text-white py-3 px-6 rounded-lg mb-4'
        >
          <Text className='text-lg text-white'>Login</Text>
        </TouchableOpacity>

        <TouchableOpacity
          onPress={handleGoogleLoginPress}
          className='items-center justify-center flex flex-row space-x-5 bg-gray-700 py-3 px-6 rounded-lg'
        >
          <Image
            className='h-6 w-6'
            source={{
              uri: "https://i.ibb.co/j82DCcR/search.png",
            }}
          />
          <Text className='text-lg text-white'>Sign in with Google</Text>
        </TouchableOpacity>
      </View>
      <View className='flex items-center mt-6'>
        <Text className='text-lg'>Forgot Password?</Text>
        <Text className='text-lg'>
          Don't have an account?
          <Text
            className='font-bold'
            onPress={() => navigation.navigate("Register")}
          >
            {" "}
            Sign Up
          </Text>
        </Text>
      </View>
    </View>
  );
};

export default LoginScreen;
