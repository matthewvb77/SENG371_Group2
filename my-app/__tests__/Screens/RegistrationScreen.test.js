import React from 'react';
import renderer from 'react-test-renderer';

import LoginScreen from '../../Screens/RegistrationScreen';

describe('<RegistrationScreen />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<LoginScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
