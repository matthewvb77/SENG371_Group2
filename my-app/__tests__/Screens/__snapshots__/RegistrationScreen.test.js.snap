// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegistrationScreen /> renders correctly 1`] = `
<View
  style={
    [
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "paddingTop": 40,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "display": "flex",
        },
        {
          "alignItems": "center",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 36,
            "lineHeight": 40,
          },
          {
            "fontWeight": "700",
          },
          {
            "marginTop": 48,
          },
        ]
      }
    >
      Register
    </Text>
  </View>
  <View
    style={
      [
        {
          "marginTop": 48,
        },
        {
          "paddingLeft": 20,
          "paddingRight": 20,
        },
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      blurOnSubmit={true}
      inputMode="email"
      keyboardType="email-address"
      onChangeText={[Function]}
      placeholder="Email"
      returnKeyType="next"
      style={
        [
          {
            "height": 48,
          },
          {
            "paddingLeft": 16,
            "paddingRight": 16,
          },
          {
            "backgroundColor": "#f3f4f6",
          },
          {
            "borderBottomWidth": 1,
            "borderLeftWidth": 1,
            "borderRightWidth": 1,
            "borderTopWidth": 1,
          },
          {
            "borderBottomColor": "#6b7280",
            "borderLeftColor": "#6b7280",
            "borderRightColor": "#6b7280",
            "borderTopColor": "#6b7280",
          },
          {
            "borderBottomLeftRadius": 8,
            "borderBottomRightRadius": 8,
            "borderTopLeftRadius": 8,
            "borderTopRightRadius": 8,
          },
          {
            "marginBottom": 16,
          },
        ]
      }
      textContentType="emailAddress"
    />
    <TextInput
      autoCapitalize="none"
      blurOnSubmit={true}
      onChangeText={[Function]}
      placeholder="Password"
      returnKeyType="next"
      secureTextEntry={true}
      style={
        [
          {
            "height": 48,
          },
          {
            "paddingLeft": 16,
            "paddingRight": 16,
          },
          {
            "backgroundColor": "#f3f4f6",
          },
          {
            "borderBottomWidth": 1,
            "borderLeftWidth": 1,
            "borderRightWidth": 1,
            "borderTopWidth": 1,
          },
          {
            "borderBottomColor": "#6b7280",
            "borderLeftColor": "#6b7280",
            "borderRightColor": "#6b7280",
            "borderTopColor": "#6b7280",
          },
          {
            "borderBottomLeftRadius": 8,
            "borderBottomRightRadius": 8,
            "borderTopLeftRadius": 8,
            "borderTopRightRadius": 8,
          },
          {
            "marginBottom": 16,
          },
        ]
      }
      textContentType="password"
    />
    <TextInput
      autoCapitalize="none"
      blurOnSubmit={true}
      onChangeText={[Function]}
      placeholder="Password"
      returnKeyType="done"
      secureTextEntry={true}
      style={
        [
          {
            "height": 48,
          },
          {
            "paddingLeft": 16,
            "paddingRight": 16,
          },
          {
            "backgroundColor": "#f3f4f6",
          },
          {
            "borderBottomWidth": 1,
            "borderLeftWidth": 1,
            "borderRightWidth": 1,
            "borderTopWidth": 1,
          },
          {
            "borderBottomColor": "#6b7280",
            "borderLeftColor": "#6b7280",
            "borderRightColor": "#6b7280",
            "borderTopColor": "#6b7280",
          },
          {
            "borderBottomLeftRadius": 8,
            "borderBottomRightRadius": 8,
            "borderTopLeftRadius": 8,
            "borderTopRightRadius": 8,
          },
          {
            "marginBottom": 16,
          },
        ]
      }
      textContentType="password"
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#3b82f6",
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "color": "#fff",
          "marginBottom": 16,
          "opacity": 1,
          "paddingBottom": 12,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 12,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontSize": 18,
              "lineHeight": 28,
            },
            {
              "color": "#fff",
            },
          ]
        }
      >
        Register
      </Text>
    </View>
  </View>
  <View
    style={
      [
        {
          "display": "flex",
        },
        {
          "alignItems": "center",
        },
        {
          "marginTop": 24,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 18,
            "lineHeight": 28,
          },
        ]
      }
    >
      Already have an account? 
    </Text>
    <Text
      onPress={[Function]}
      style={
        [
          {
            "fontWeight": "700",
          },
          {
            "fontSize": 18,
            "lineHeight": 28,
          },
        ]
      }
    >
      Login
    </Text>
  </View>
</View>
`;
